### ISSUE TITLE
JWKS using Jose2 doesnt work anymore with new versions of Oidc Provider

### ISSUE THREAD

**Issue by MeydanOzeri** [panva/node-oidc-provider#960]
Oidc Provider was updated to support Jose v3, but the documentations and examples of how to provide the JWKS are outdated to V2, and now when updating the Oidc Provider to the new version they dont work anymore.

this is the error i get:

Error: jwks.keys[0] is of no use given the other configuration, remove it

It would be nice to give an example of how to provide JWKS with Jose V3 or to still support Jose V2.

This was tested using oidc-provider v7.1.3 and jose v2.0.3.

**Comment by panva** [comment]
Your "issue" is completely unrelated to the jose version used. It's just a new validation in place. Basically the key is of no use to the OP given its configuration. 

Feel free to provide the configuration and key in question just for me to confirm.

**Comment by vanthome** [comment]
I have an issue that seems to be related. With the latest version `7.1.3` of `oidc-provider`:

```
Cannot find module 'jose/jwk/parse' from 'node_modules/oidc-provider/lib/helpers/keystore.js
```

**Comment by vanthome** [comment]
@panva can you please comment on my issue or should I open a new one for this?

**Comment by panva** [comment]
> @panva can you please comment on my issue or should I open a new one for this?

re-install your dependencies? ensure your node.js version is the one required by v7.x? basic stuff, no need to open an issue as i won't be able to help with local setup issues.

**Comment by MeydanOzeri** [comment]
> Your "issue" is completely unrelated to the jose version used. The format is also agnostic to jose, because its simply a JWKSet formatted javascript object. It's just a new validation in place. Basically the key is of no use to the OP given its configuration.
> 
> Feel free to provide the configuration and key in question just for me to confirm.

Clients:

```js
const externalUserOidcProviderClient = () => {
	const { EXTERNAL_USER_API_NAME, EXTERNAL_USER_API_PASSWORD, CLIENT_URL } = process.env;
	return {
		client_id: EXTERNAL_USER_API_NAME,
		client_secret: EXTERNAL_USER_API_PASSWORD,
		grant_types: ['authorization_code', 'client_credentials'],
		redirect_uris: [CLIENT_URL]
	};
};

const webAppOidcProviderClient = () => {
	const { API_NAME, API_PASSWORD, CLIENT_URL } = process.env;
	return {
		application_type: 'web',
		client_id: API_NAME,
		client_secret: API_PASSWORD,
		grant_types: ['refresh_token', 'authorization_code'],
		response_types: ['code'],
		scope: 'openid offline_access',
		redirect_uris: [`${CLIENT_URL}/login`],
		post_logout_redirect_uris: [`${CLIENT_URL}/login`]
	};
};

const initOidcProviderClients = () => [externalUserOidcProviderClient(), webAppOidcProviderClient()];

export { initOidcProviderClients };
```

Config: 
```js
import fs from 'fs';
import ms from 'ms';
import OidcProvider from 'oidc-provider';
import _ from 'lodash';
import { initOidcProviderClients } from './oidcProviderClients.js';
import { files } from '../../../shared/index.js';

const { interactionPolicy, Provider } = OidcProvider;

const { SERVER_HOST } = process.env;

const oidcProviderPath = files.getFilePath('/src/services/auth/oidcProvider');

const loadJwks = () => JSON.parse(fs.readFileSync(`${oidcProviderPath}/jwks.json`).toString());

const loadCookiesKeys = () => JSON.parse(fs.readFileSync(`${oidcProviderPath}/cookiesKeys.json`).toString());

const initOidcProvider = ({ providerServerUrl, adapter, findAccount }) => {
	const policy = interactionPolicy.base();
	const providerConfiguration = {
		clients: initOidcProviderClients(),
		adapter,
		findAccount,
		formats: { AccessToken: 'jwt' },
		jwks: loadJwks(),
		cookies: {
			long: { signed: true, httpOnly: true, secure: _.includes(SERVER_HOST, 'https'), overwrite: true, maxAge: ms('1d') },
			short: { signed: true, httpOnly: true, secure: _.includes(SERVER_HOST, 'https'), overwrite: true, maxAge: ms('1d') },
			keys: loadCookiesKeys()
		},
		pkce: { methods: ['S256'], required: () => true },
		responseTypes: ['code'],
		interactions: { policy, url: (ctx) => `/interaction/${ctx.oidc.uid}` },
		scopes: ['openid', 'offline_access'],
		claims: { openid: ['userType', 'userId', 'username', 'userFingerprint'] },
		extraParams: ['userFingerprint'],
		expiresWithSession: () => true,
		features: {
			devInteractions: { enabled: false },
			revocation: { enabled: true },
			clientCredentials: { enabled: true },
			rpInitiatedLogout: {
				enabled: true,
				logoutSource: async (ctx, form) => {
					const confirmLogoutButton = '<button id="logout-button" hidden type="submit" form="op.logoutForm" value="yes" name="logout"></button>';
					const clickLogoutButtonScript = `<script nonce="${ctx.req.app.nonce}">document.getElementById("logout-button").click()</script>`;
					ctx.body = `<!DOCTYPE html>${form}${confirmLogoutButton}${clickLogoutButtonScript}</html>`;
				}
			}
		},
		ttl: {
			AccessToken: ms('1h') / 1000,
			AuthorizationCode: ms('10m') / 1000,
			ClientCredentials: ms('100y') / 1000,
			DeviceCode: ms('1h') / 1000,
			IdToken: ms('1h') / 1000,
			RefreshToken: ms('1d') / 1000
		}
	};
	const oidcProvider = new Provider(providerServerUrl, providerConfiguration);
	oidcProvider.proxy = true;
	return oidcProvider;
};

export { initOidcProvider };
```

**Comment by panva** [comment]
And the jwks?

**Comment by MeydanOzeri** [comment]
> And the jwks?

This would be the function that generates them.

```js
const createAuthKeys = async () => {
	const keyStore = new jose.JWKS.KeyStore();
	await Promise.all([
		keyStore.generate('RSA', 2048, { use: 'sig' }),
		keyStore.generate('RSA', 2048, { use: 'enc' }),
		keyStore.generate('EC', 'P-256', { use: 'sig' }),
		keyStore.generate('EC', 'P-256', { use: 'enc' }),
		keyStore.generate('OKP', 'Ed25519', { use: 'sig' })
	]);
	return { jwks: JSON.stringify(keyStore.toJWKS(true), null, 2), cookiesKeys: JSON.stringify([OpenidClient.generators.random(64)]) };
};
```

**Comment by panva** [comment]
Okay, please re-read what I said. You're generating keys that are of no use the provider because of its configuration. In your case - you're adding `use: enc` encryption keys but have the encryption feature disabled.

Bottom line everything's in the validation error message you're getting.

**Comment by MeydanOzeri** [comment]
> 
> 
> Okay, please re-read what I said. You're generating keys that are of no use the provider because of its configuration. In your case - you're adding `use: enc` encryption keys but have the encryption feature disabled.
> 
> Bottom line everything's in the validation error message you're getting.

Ok it worked fine thank you for the guidance, but im experiencing another issue now.
Im able to load the login screen, but it fails after authentication. (it doesnt move the user to the consent screen)
After enabling the debug i found this error: TypeError: Cannot read property 'getOIDCScopeEncountered' of undefined.
I tried to find in the changelog of v7 what was changed related to this, and i saw some lines talking about consent and grant changes, but i cant find any example or explanation what and where i need to change something about the grant.
Help would be appreciated.
